### ISSUE TITLE
Highlight that `int | float` or `Union[int, float]` in type annotations simplify to `float`

### ISSUE THREAD

**Issue by gpshead** [PyCQA/pylint#7084]
### Current problem

Given:

```python
from __future__ import annotations
from typing import Union

def function(arg: Union[int, float]):
    return float(arg)

def function2(arg: int | float):
    return float(arg)
```

Those annotations can logically be shortened to `float` thanks to the numeric tower and as mentioned in PEPs 483 and 484. Taking this further the thing to suggest is just that int be removed from the union list when float is present. Which in this degenerate but common example happens to mean there is no more union.

ie: `int | float | bytes` would become `float | bytes`.

### Desired solution

The type annotation in the code is logically correct, but _could_ be made more concise. I suggest a low importance lint warning for this case. Perhaps suggest that "float in a type union already implies int, consider removing int" or somesuch wording if you don't want to get into a long winded explanation.

### Additional context

https://peps.python.org/pep-0484/ and https://peps.python.org/pep-0483/

**Comment by cdce8p** [comment]
This would probably be much easier to implement in `mypy`. Pylint is only a linter, not a type checker. As such any solution here would be quite difficult to get right and hard to maintain in the long term.

**Comment by Pierre-Sassoulas** [comment]
I think it's only a localized check on a very specific case (When we have a typing union with int and float), why would it be hard to get right, Marc ?

**Comment by cdce8p** [comment]
> I think it's only a localized check on a very specific case (When we have a typing union with int and float), why would it be hard to get right, Marc ?

It starts out easy. I was just thinking a few steps ahead to a more generalized solution. E.g. what about subclasses or type aliases? Should pylint warn about them, too? From my experience anything type related is much better done by an actual type checker. _I even suspect this could be fairly simple to implement in mypy._

**Comment by cdce8p** [comment]
If it were a runtime issue, I would consider it. However, as it's only a nice to have with no impact on the code itself, I would argue that it's not worth the effort to implement it here.

**Comment by Pierre-Sassoulas** [comment]
Let's wait for some other opinions before closing as won't do (or for the release of 2.15).

**Comment by gpshead** [comment]
There are ~5+ different Python type checkers. There are ~2 Python linters. So I proposed it here because this _felt_ like more of a Style issue than a correctness issue. Closing it as "won't do" or "only if someone pops up to author the checker PR" is fine if desired. This was just a thought based on an internal style / code readability discussion to see if anyone else was interested.

As a lint check it could even just look for the specific exact int + float and nothing else union pattern, it shouldn't attempt full type intersection analysis. I see `Union[int, float]` appear in our internal code base >1000 times (but given its >100m line size, that is not a huge number).

As for type checkers, I'm personally used to pytype, which is in the business of reporting hard errors. I don't know if Python type checkers in general have a non-fatal warning concept and are integrated into development workflows in such a way that developers would actually _see_ those warnings. _(an issue for type checkers and IDE integrations to work out)_

**Comment by cdce8p** [comment]
> This was just a thought based on an internal style / code readability discussion to see if anyone else was interested.

A check for it (and especially the more general case) would undoubtably be useful. Pylint isn't the best tool for it though, unfortunately.

> As a lint check it could even just look for the specific exact int + float and nothing else union pattern, it shouldn't attempt full type intersection analysis. I see `Union[int, float]` appear in our internal code base >1000 times (but given its >100m line size, that is not a huge number).

It won't stay limited for long. Other users will request their edge cases and before you know it, you're doing a complete type analysis.

> I don't know if Python type checkers in general have a non-fatal warning concept and are integrated into development workflows in such a way that developers would actually _see_ those warnings.

I mainly use mypy and pyright though VS Code and pylance. Both I think would implement / accept such a check. One obstacle could be performance though. Since the subclass relationship is directional, you would need two calls per type pair. That can get expensive fast for unions with a lot of types.
